File: fetcher/src/main/java/com/freehand/fetcher/FetchBag.java
package com.freehand.fetcher;

import java.util.HashSet;
import java.util.Set;

/**
 * Created by minhpham on 12/9/18.
 * Purpose: control multiple fetcher
 * Copyright © 2018 Pham Duy Minh. All rights reserved.
 */
public class FetchBag {

    private Set<IFetcher> bag;

    /**
     * add fetcher
     *
     * @param fetcher
     */
    public void addFetcher(IFetcher fetcher) {
        if(fetcher == null) return;
        if (bag == null) {
            bag = new HashSet<>();
        }
        bag.add(fetcher);
    }

    /**
     * cancel all fetcher
     */
    public void cancel() {
        if (bag == null) return;
        for (IFetcher fetcher : bag) {
            fetcher.cancel();
        }
    }

    /**
     * destroy all fetcher
     */
    public void destroy() {
        if (bag == null) return;
        for (IFetcher fetcher : bag) {
            fetcher.destroy();
        }
        bag.clear();
        bag = null;
    }
}

File: fetcher/src/main/java/com/freehand/fetcher/FetchBag.kt
package com.freehand.fetcher

import java.util.HashSet

/**
 * Created by minhpham on 12/9/18.
 * Purpose: control multiple fetcher
 * Copyright © 2018 Pham Duy Minh. All rights reserved.
 */
class FetchBag {

    private var bag: MutableSet<IFetcher<*, *>>? = null

    /**
     * add fetcher
     *
     * @param fetcher
     */
    fun addFetcher(fetcher: IFetcher<*, *>?) {
        if (fetcher == null) return
        if (bag == null) {
            bag = HashSet()
        }
        bag?.add(fetcher)
    }

    /**
     * cancel all fetcher
     */
    fun cancel() {
        if (bag == null) return
        for (fetcher in bag!!) {
            fetcher.cancel()
        }
    }

    /**
     * destroy all fetcher
     */
    fun destroy() {
        if (bag == null) return
        for (fetcher in bag!!) {
            fetcher.destroy()
        }
        bag?.clear()
        bag = null
    }
}
